core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2914,6 +2928,108 @@ public void testSequenceIDs() throws Exception {
         searchResult.close();
     }
 
+    public void testSequenceNumberAdvancesToMaxSeqNoOnEngineOpen() throws IOException {
+        engine.close();
+        final long maxSeqNo = randomIntBetween(16, 32);
+        final long localCheckpoint =
+            rarely() ? SequenceNumbersService.NO_OPS_PERFORMED : randomIntBetween(0, Math.toIntExact(maxSeqNo));
+        final long globalCheckpoint =
+            localCheckpoint == SequenceNumbersService.NO_OPS_PERFORMED || rarely()
+                ? SequenceNumbersService.UNASSIGNED_SEQ_NO
+                : Math.toIntExact(localCheckpoint);
+        Engine initialEngine = null;
+        try {
+            initialEngine = createEngine(
+                defaultSettings,
+                store,
+                primaryTranslogDir,
+                newMergePolicy(),
+                null,
+                () -> new SequenceNumbersService(shardId, defaultSettings, maxSeqNo, localCheckpoint, globalCheckpoint) {
+                    @Override
+                    public long generateSeqNo() {
+                        if (rarely()) {
+                            // force skipping a sequence number
```

Comment:
I wonder if we should do the test based on real world scenarios we know can happen. I think it will make things clearer?

For example, using a replica engine :
1) Index a doc, with seq#1
2) Delete the same doc with seq#3
3) Flush
4) index doc with seq#2
 
To do so on a primary takes something more evil - instead of skiping translog ops, make sure this come "out of order". So make the seq number service put a seq# aside but then serve it on a follow up request.

I think this also serve as documentation of when this can happen.